syntax = "proto3";

package dicer;

// When dicer is run with the -C option, it generates complementary
// fragments. those have the attachment points labelled with
// the complementary smiles.
message Fragment {
  // the fragment generated by dicer.
  string fragment = 1;

  // The complementary fragments.
  string complement = 2;

  // Within `complement` there will be isotopic atoms.
  // This is a mapping from those isotopic labels to
  // atom types.
  map<int32, uint32> atype = 3;
}

message DicedMolecule {
  // The parent molecule.
  string parent_smiles = 1;
  // The id of the parent molecule.
  string parent_id = 2;

  // There will be an arbitrary number of fragments.
  repeated Fragment fragment = 3;
}

message DicerFragment {
  // The smiles of the fragment. Will contain isotopically
  // labeled atoms for the join points.
  string smiles = 1;

  // The first molecule that instantiated this fragment.
  string name = 2;

  // This is a mapping from those isotopic labels to
  // atom types.
  // map<int, uint32> atype = 3;
  string atype = 3;

  // It is convenient to keep the inter-isotopic distances here.
  // Not necessary because most likely they will be used as a key
  // but convenient.
  // The convention is that ordering is d12 d13 d14 ... d23 d24...
  repeated uint32 dist = 4;

  // The number of molecules exemplifying this fragment.
  uint32 n = 5;
}

message SmilesIdCount {
  string smiles = 1;
  string id = 2;
  string atype = 3;
  uint32 n = 4;
}
